File: frontend/src/services/admin/adminUserService.ts
import apiClient, { type ApiResponse } from '../api';
import type { PaginatedResponse } from '../productService';

export interface UserResponse {
  id: number;
  email: string;
  firstName: string;
  lastName: string;
  phone: string;
  role: 'USER' | 'ADMIN';
  isVerified: boolean;
  isActive: boolean;
  createdAt: string;
}

export interface UpdateUserRequest {
  firstName: string;
  lastName: string;
  phone: string;
}

export const AdminUserService = {
  getUsers: async (params?: {
    query?: string;
    isActive?: boolean;
    role?: 'USER' | 'ADMIN';
    specificMonth?: string; // format yyyy-MM
    page?: number;
    size?: number;
  }) => {
    const response = await apiClient.get<ApiResponse<PaginatedResponse<UserResponse>>>('/admin/users', { params });
    // @ts-ignore
    return response.data as PaginatedResponse<UserResponse>;
  },

  updateUserStatus: async (id: number, isActive: boolean) => {
    const response = await apiClient.patch<ApiResponse<UserResponse>>(`/admin/users/${id}/status`, null, {
      params: { isActive }
    });
    return response.data;
  },

  updateUserRole: async (id: number, role: 'USER' | 'ADMIN') => {
    const response = await apiClient.patch<ApiResponse<UserResponse>>(`/admin/users/${id}/role`, null, {
      params: { role }
    });
    return response.data;
  },

  updateUserInfo: async (id: number, request: UpdateUserRequest) => {
    const response = await apiClient.put<ApiResponse<UserResponse>>(`/admin/users/${id}`, request);
    return response.data;
  }
};

File: backend/src/main/java/com/peaksneaker/service/AdminUserService.java
package com.peaksneaker.service;

import com.peaksneaker.dto.request.UpdateUserRequest;
import com.peaksneaker.dto.response.PaginatedResponse;
import com.peaksneaker.dto.response.UserResponse;
import com.peaksneaker.entity.User;
import com.peaksneaker.enums.Role;
import com.peaksneaker.repository.UserRepository;
import com.peaksneaker.repository.specification.UserSpecification;
import lombok.RequiredArgsConstructor;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.Instant;

@Service
@RequiredArgsConstructor
@Transactional(readOnly = true)
public class AdminUserService {

    private final UserRepository userRepository;

    public PaginatedResponse<UserResponse> getUsers(
            String query, Boolean isActive, Role role, Instant startDate, Instant endDate, int page, int size
    ) {
        Pageable pageable = PageRequest.of(page, size, Sort.by(Sort.Direction.DESC, "createdAt"));
        Specification<User> spec = UserSpecification.filterUsers(query, isActive, role, startDate, endDate);

        Page<User> userPage = userRepository.findAll(spec, pageable);
        Page<UserResponse> responsePage = userPage.map(this::mapToResponse);

        return PaginatedResponse.from(responsePage);
    }

    @Transactional
    public UserResponse updateUserStatus(Long id, Boolean isActive) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Không tìm thấy người dùng với ID: " + id));

        if (isActive) {
            user.activate();
        } else {
            user.block();
        }

        User savedUser = userRepository.save(user);
        return mapToResponse(savedUser);
    }

    @Transactional
    public UserResponse updateUserRole(Long id, Role role) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Không tìm thấy người dùng với ID: " + id));

        user.setRole(role);
        User savedUser = userRepository.save(user);
        return mapToResponse(savedUser);
    }

    @Transactional
    public UserResponse updateUserInfo(Long id, UpdateUserRequest request) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new IllegalArgumentException("Không tìm thấy người dùng với ID: " + id));

        user.updateProfile(request.getFirstName(), request.getLastName(), request.getPhone());
        User savedUser = userRepository.save(user);
        return mapToResponse(savedUser);
    }

    private UserResponse mapToResponse(User user) {
        return UserResponse.builder()
                .id(user.getId())
                .email(user.getEmail())
                .firstName(user.getFirstName())
                .lastName(user.getLastName())
                .fullName(user.getFullname())
                .phone(user.getPhone())
                .role(user.getRole())
                .isVerified(user.getIsVerified())
                .isActive(user.getIsActive())
                .createdAt(user.getCreatedAt())
                .build();
    }
}
